cds_ils/migrator/series/api.py

```diff
@@ -66,75 +71,122 @@ def create_multipart_volumes(pid, multipart_legacy_recid, migration_volumes):
+            # the barcode represents an item
+            if "items" not in volume:
+                volume["items"] = []
+            volume["items"].append({
+                "barcode": obj["barcode"]
+            })
+        else:
+            # all other fields should be treated as
+            # additional metadata for the document
+            for key in obj:
+                if key != "volume":
+                    if key in volume:
+                        # abort in case of conflict
+                        raise KeyError(
+                            'Duplicate key "{}" for multipart {}'.format(
+                                key, multipart_legacy_recid
+                            )
                         )
-                    )
-                volume[key] = obj[key]
+                    volume[key] = obj[key]
 
     volume_numbers = iter(sorted(volumes.keys()))
 
-    # Re-use the current record for the first volume
-    # TODO review this - there are more cases of multiparts
-    first_volume = next(volume_numbers)
-    first = Document.get_record_by_pid(pid)
-    if "title" in volumes[first_volume]:
-        first["title"] = volumes[first_volume]["title"]
-        first["volume"] = first_volume
-    first["_migration"]["multipart_legacy_recid"] = multipart_legacy_recid
+    inherited_metadata = deepcopy(document_base_metadata)
+    inherited_metadata["_migration"]["multipart_legacy_recid"] = \
+        multipart_legacy_recid
+    inherited_metadata["authors"] = \
+        inherited_metadata["_migration"]["authors"] \
+        if "authors" in inherited_metadata["_migration"] else []
+    inherited_metadata["serial_title"] = inherited_metadata.get("title")
+
     # to be tested
-    if "legacy_recid" in first:
-        del first["legacy_recid"]
-    first.commit()
-    yield first
+    if "legacy_recid" in inherited_metadata:
+        del inherited_metadata["legacy_recid"]
 
     # Create new records for the rest
     for number in volume_numbers:
-        temp = first.copy()
-        temp["title"] = volumes[number]["title"]
+        volume = volumes[number]
+        temp = inherited_metadata.copy()
+        if "title" in volume and volume["title"]:
+            temp["title"] = volume["title"]
         temp["volume"] = number
+        # TODO possibly more fields to merge
+
         record_uuid = uuid.uuid4()
-        provider = DocumentIdProvider.create(
-            object_type="rec", object_uuid=record_uuid
-        )
-        temp["pid"] = provider.pid.pid_value
-        record = Document.create(temp, record_uuid)
-        record.commit()
-        yield record
+        try:
+            with db.session.begin_nested():
+                provider = DocumentIdProvider.create(
+                    object_type="rec", object_uuid=record_uuid
+                )
+                temp["pid"] = provider.pid.pid_value
+                record = Document.create(temp, record_uuid)
+                record.commit()
```

Comment:
I think we don't need the `record.commit()` here, it is already done with the `create`, otherwise it will commit 2 times.